src/database/Database.cpp

```diff
@@ -289,6 +302,202 @@ Database::upgradeToCurrentSchema()
+                                  column);
+        CLOG(INFO, "Database")
+            << "Dropping column '" << column << "' with string '"
+            << dropColumnStr << "' from table '" << table << "'";
+
+        mSession << dropColumnStr;
+    }
+    else
+    {
+        CLOG(INFO, "Database") << "SQLite does not support dropping column '"
+                               << column << "' from table '" << table << "'";
+    }
+}
+
+StatementContext
+Database::getPreparedOldLiabilitySelect(std::string const table,
+                                        std::string const fields)
+{
+    return getPreparedStatement("SELECT " + fields + "," +
+                                "buyingliabilities, sellingliabilities "
+                                "FROM " +
+                                table +
+                                " WHERE "
+                                "buyingliabilities IS NOT NULL"
+                                " OR "
+                                "sellingliabilities IS NOT NULL");
+}
+
+void
+Database::convertAccountExtensionsToOpaqueXDR()
+{
+    addTextColumnIfNotPresent("accounts", "extension");
+    copyIndividualAccountExtensionFieldsToOpaqueXDR();
+    dropTextColumn("accounts", "buyingliabilities");
+    dropTextColumn("accounts", "sellingliabilities");
+}
+
+void
+Database::convertTrustLineExtensionsToOpaqueXDR()
+{
+    addTextColumnIfNotPresent("trustlines", "extension");
+    copyIndividualTrustLineExtensionFieldsToOpaqueXDR();
+    dropTextColumn("trustlines", "buyingliabilities");
+    dropTextColumn("trustlines", "sellingliabilities");
+}
+
+void
+Database::copyIndividualAccountExtensionFieldsToOpaqueXDR()
+{
+    CLOG(INFO, "Ledger") << __func__ << ": updating account extension schema";
+
+    std::string accountIDStrKey;
+    AccountEntry::_ext_t::_v1_t extension;
+    soci::indicator buyingLiabilitiesInd, sellingLiabilitiesInd;
+
+    auto prep_select = getPreparedOldLiabilitySelect("accounts", "accountid");
+    auto& st_select = prep_select.statement();
+    st_select.exchange(soci::into(accountIDStrKey));
+    st_select.exchange(
+        soci::into(extension.liabilities.buying, buyingLiabilitiesInd));
+    st_select.exchange(
+        soci::into(extension.liabilities.selling, sellingLiabilitiesInd));
+    st_select.define_and_bind();
+    {
+        auto timer = getSelectTimer("account-ext-to-opaque");
+        st_select.execute(true);
+    }
+
+    size_t numAccountsUpdated = 0;
+    for (; st_select.got_data(); st_select.fetch())
+    {
+        // We've only selected accounts which have at least one of
+        // buying liabilities or selling liabilities present, and if
+        // either is present, then both should be.
+        assert(buyingLiabilitiesInd == soci::i_ok);
+        assert(sellingLiabilitiesInd == soci::i_ok);
+        std::string opaqueExtension(
+            decoder::encode_b64(xdr::xdr_to_opaque(extension)));
+        auto prep_update = getPreparedStatement(
```

Comment:
Ah, I didn't realize it was possible to reuse a prepared statement.  I've now done that.

However, having done so, I'm now not sure about it.  If I understand what that's doing correctly, it's what this page refers to as "Statement caching":

http://soci.sourceforge.net/doc/master/statements/

And that page says that to reuse the prepared statement, we should follow the ```execute()``` call with a call to ```bind_clean_up()```.  But I can't find a ```bind_clean_up()``` in the version of soci that we're using.

As for the interleaving, I wondered about that myself, thought I'd convinced myself that it was safe, and after your question am now unsure again.  Perhaps it's easier to confirm that we'll be able to hold all the changes in memory for the real database than it is to confirm that this code is always safe (now that my Googling has also failed to produce a definitive answer).  I'll change that and experiment with it.

Reply:
I have now changed the upgrade code to separate the select from the update rather than interleaving them.  While I was at it, I also factored out the commonality between the account and trustline updates, so that there's one place (a templated function called ```copyIndividualExtensionFieldsToOpaqueXDR()```) where we can experiment with the algorithm.